README.md
# **highrise.sdk**
<p align="center">
  <img src="https://i.ibb.co/d0vtV49/highrise-logo.png" alt="highrise-logo" />
</p>

> **The Highrise SDK is a JavaScript library for writing and running Highrise bots.**


## **⚙️ Installation** 
```
npm i highrise.sdk@latest
```

## **✨ Features**

- Easy to use.
- Beginner friendly.
- Auto reconnect system.
- Supports Node version 10+
- Supports Highrise WebAPI

## **📥 Class Import**
1. Import the necessary classes and modules from the SDK:
```js
const { Highrise, GatewayIntentBits, WebApi } = require('highrise.sdk');
```

2. Set up the bot's settings by providing a bot token and room ID:
```js
const settings = {
  token: 'CHANGE-ME', // Replace with your bot token
  room: 'CHANGE-ME', // Replace with the room ID your bot will join
  reconnect: 5 // Reconnect duration in seconds
};
```

3. Create an instance of the Highrise bot, specifying the desired intents and cache option:
```js
const bot = new Highrise({
  intents: [
    GatewayIntentBits.Ready,
    GatewayIntentBits.Messages,
    GatewayIntentBits.Joins,
    GatewayIntentBits.Leaves,
    GatewayIntentBits.Error
  ],
  cache: true
}, settings.reconnect);
```

3. Logging in the bot:
```js
bot.login(settings.token, settings.room);
```

## **🎋 GatewayIntentBits**
GatewayIntentBits represents the different intents or event types that your bot can listen for. By specifying these intents when creating the bot, you can control which events your bot will receive. The available intents include:

- `GatewayIntentBits.Ready`: Indicates when the bot is ready to start interacting.
- `GatewayIntentBits.Messages`: Represents chat messages sent in the Highrise room.
- `GatewayIntentBits.DirectMessages`: Represents direct messages sent to the bot.
- `GatewayIntentBits.Joins`: Indicates when users join the room.
- `GatewayIntentBits.Leaves`: Indicates when users leave the room.
- `GatewayIntentBits.Reactions`: Represents reactions added to players.
- `GatewayIntentBits.Emotes`: Represents emotes added to players.
- `GatewayIntentBits.Tips`: Represents tip reactions received.
- `GatewayIntentBits.VoiceChat`: Represents voice chat events.
- `GatewayIntentBits.Movements`: Indicates when users move within the room.
- `GatewayIntentBits.Error`: Represents errors that occur during API operations.
- `GatewayIntentBits.Moderate`: Indicates when moderators perform moderation actions on players.

You can choose the intents based on the events you want your bot to handle.

## **📦 Cache Option**
The cache option, when set to true, enables caching of certain data to optimize performance and reduce API calls. By enabling the cache and using the appropriate intents, you can utilize methods that rely on cached data rather than making API requests.

## **📖 Examples**
- Listening for the ready event:
```js
bot.on('ready', (session) => {
  // Handle bot ready event
});
```
- Listening for chat messages:
```js
bot.on('chatCreate', async (user, message) => {
  // Handle chat message event
});
```
- Listening for errors:
```js
bot.on('error', (message) => {
  // Handle error event
});
```

## **📘 Documentation**
Refer to the SDK documentation for more information on available events and methods.
[Highrise JS SDK Documentation](https://bit.ly/highrise-sdk)

## **🤝 Contributions**
Contributions to the Highrise SDK are welcome! If you find any issues or want to add new features, feel free to submit a pull request.

## Note

This package is not an official Highrise package, it's self-made by iHsein (sphinix) and is still in beta.



CHANGELOG.md
# Changelog

## [1.1.7] - 2023-09-29
## Fixed
- Fixed an issue where, upon shutting down the bot, the websocket would sometimes remain open.

## [1.1.6] - 2023-08-10
## Added
- Introducing a new event that emits after performing moderation actions ["kick", "ban", "unban", "mute", "unmute"].
- Implementation of a new Gateway for the "Moderate" event.
- Addition of a new method that enables you to gracefully shut down the bot.

## Fixed
- Addressed issues with WebApi, where it would occasionally return an error when no results were found. It has been enhanced to now return "null" in such cases.

## [1.1.5] - 2023-08-02
## Added
- A new parameter to the "playerJoin" event that returns the player's position.
- playerTip event now work in both chat and inbox for more information [read this](https://createforum.highrise.game/t/python-sdk-version-23-3-1-bot-api-update/152)
- A new method to fetch the bot information (id, position).

## Changed
- Returning the error data is no longer available. The package should not log the error; instead, it will log "reconnecting."
- You can now retrieve the bot's id and position by using "bot.client.id.get" or "bot.client.position.get." This will be updated automatically when calling any request that includes the bot.

## Fixed
- Caching the player's data will now cache the player's position on join. Instead of 0,0,0, it will cache the actual position.
- Using "bot.move" or "bot.player.teleport" will now save the bot's position.

## [1.1.4] - 2023-07-17
## Added
- Implemented new bot customization methods.
- Expanded the WebApi to retrieve item and grab data.

## Changed
- Reorganized the code structure by merging the WalletRequest file into the BotRequest file to handle all bot requests in one place.

## Fixed
- Resolved issues with the WebApi methods that were causing crashes in the bot.

## [1.1.3] - 2023-07-08
## Fixed
- Applied a hotfix to the WebApi, resolving the issue where some methods were returning ERROR 404.

## [1.1.1] - 2023-07-05
## Fixed
- A hotfix has been applied to the Highrise class, resolving the issue where importing the class resulted in a "TypeError: Highrise is not a constructor" error.


## [1.1.0] - 2023-07-05
## Added
- New tip methods: Now the bot can tip any player using the `player.tip(user_id, amount)` function. This function returns a promise that resolves to either "success" or "insufficient_funds". The available tip amounts are similar to the ones in the game: "1, 5, 10, 50, 100, 500, 1000, 5000, 10000". Please note that similar to player tipping, a fee is applied.

## Fixed
- Fixed dot notation issue for methods.
- Fixed a crash caused by the getNewsFeed function in the WebApi when attempting to fetch a post.

## [1.0.0]

## [1.0.0] - 2023-07-04
## Added
- New WebAPI support.
- Introducing the option to enable or disable cache.
- Expanded the available intents to provide more event handling capabilities.


## Changed
- Consolidated the previous [highrise-js-sdk](https://github.com/sphinixFTW/highrise-js-sdk) package into the current [highrise.sdk](https://github.com/sphinixFTW/highrise.sdk) package. The old package will no longer be supported as we focus on testing and improving the new version.
- Restructured the folder organization for improved readability and accessibility.
- Implemented a new event handling system using the newly introduced intents, allowing for more streamlined event management within the package.
- Modified the behavior of the cache functionality, now providing the ability to enable or disable it as needed.
- Enhanced the reconnect and connect system to improve stability and reliability.
- Standardized event names across the package for consistency. For example, the event `chatMessageCreate` has been renamed to `chatCreate`. Similar changes have been made to other events. Refer to the [documentation](bit.ly/highrise-sdk) for more information.

src/actions/UserRequests.js
const { HighriseError } = require("../handlers/error");
const { InviteSpeakerRequest, RemoveSpeakerRequest, ModerateRoomRequest, EmoteRequest, MoveUserToRoomRequest, TeleportRequest, Position, SendPayloadAndGetResponse, GetUserOutfitRequest, GetRoomPrivilegeRequest, GetBackpackRequest, ReactionRequest, TipUserRequest } = require("../models/models");
const { generateRid, CachedRoomUsers, getBotInfo, updateBotPosition } = require("../utils/Utils");

class Users {
  constructor(bot) {
    this.bot = bot;
    this.rid = generateRid();
  }

  voice = {
    add: async (user_id) => {
      try {
        if (!user_id || typeof user_id !== 'string') {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }

        // Create an instance of InviteSpeakerRequest
        const inviteRequest = new InviteSpeakerRequest(user_id);

        // Create the payload using the request object
        const payload = {
          _type: 'InviteSpeakerRequest',
          ...inviteRequest
        };

        // Send the invite request
        if (this.bot.ws.readyState === this.bot.websocket.OPEN) {
          this.bot.ws.send(JSON.stringify(payload), (error) => {
            if (error) {
              throw new HighriseError("Error sending InviteSpeakerRequest:".red);
            }
          });
        }

      } catch (error) {
        console.error(error);
      }
    },
    remove: async (user_id) => {
      try {
        if (!user_id || typeof user_id !== 'string') {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }

        // Create an instance of RemoveSpeakerRequest
        const removeRequest = new RemoveSpeakerRequest(user_id);

        // Create the payload using the request object
        const payload = {
          _type: 'RemoveSpeakerRequest',
          ...removeRequest
        };

        // Send the remove request
        if (this.bot.ws.readyState === this.bot.websocket.OPEN) {
          this.bot.ws.send(JSON.stringify(payload), (error) => {
            if (error) {
              throw new HighriseError("Error sending RemoveSpeakerRequest:".red);
            }
          });
        }

      } catch (error) {
        console.error(error);
      }
    }
  };

  async moderateRoom(request) {
    try {
      const payload = {
        _type: "ModerateRoomRequest",
        user_id: request.user_id,
        moderation_action: request.moderation_action,
        action_length: request.action_length !== undefined ? request.action_length : null,
        rid: this.rid
      };

      return new Promise((resolve, reject) => {
        if (this.bot.ws && this.bot.ws.readyState === this.bot.websocket.OPEN) {
          this.bot.ws.send(JSON.stringify(payload), error => {
            if (error) {
              console.error("Error sending ModerateRoomRequest:".red, error);
              reject(error);
              throw new HighriseError("Error sending ModerateRoomRequest:".red)
            } else {
              resolve();
            }
          });
        } else {
          reject();
        }
      });
    } catch (error) {
      console.error(error)
    }
  };

  async kick(user_id) {
    try {
      if (!user_id || typeof user_id !== 'string') {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }
      await this.moderateRoom(
        new ModerateRoomRequest(user_id, "kick")
      );
    } catch (error) {
      console.log(error)
    }
  };

  async ban(user_id, seconds) {
    try {
      if (!user_id || typeof user_id !== 'string') {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }
      if (!Number.isInteger(seconds)) {
        throw new HighriseError('Invalid duration. Duration must be an integer.'.red);
      }
      await this.moderateRoom(
        new ModerateRoomRequest(user_id, "ban", seconds)
      );
    } catch (error) {
      console.log(error)
    }
  };

  async mute(user_id, seconds) {
    try {
      if (!user_id || typeof user_id !== 'string') {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }
      if (!Number.isInteger(seconds)) {
        throw new HighriseError('Invalid duration. Duration must be an integer.'.red);
      }
      await this.moderateRoom(
        new ModerateRoomRequest(user_id, "mute", seconds)
      );
    } catch (error) {
      console.log(error)
    }
  };

  async unmute(user_id) {
    try {
      if (!user_id || typeof user_id !== 'string') {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }
      await this.moderateRoom(
        new ModerateRoomRequest(user_id, "mute", 1)
      );
    } catch (error) {
      console.log(error)
    }
  };

  async unban(user_id) {
    try {
      if (!user_id || typeof user_id !== 'string') {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }
      await this.moderateRoom(
        new ModerateRoomRequest(user_id, "unban", null)
      );
    } catch (error) {
      console.log(error)
    }
  };

  async emote(user_id, emote_id) {
    try {
      if (!emote_id || typeof emote_id !== 'string') {
        throw new HighriseError('Invalid emote ID. Please provide a valid string for emote ID.');
      }

      if (!user_id || typeof user_id !== 'string') {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }

      const emoteRequest = new EmoteRequest(user_id, emote_id, this.rid);

      const payload = {
        _type: 'EmoteRequest',
        emote_id: emoteRequest.emote_id,
        target_user_id: emoteRequest.target_user_id,
        rid: emoteRequest.rid
      };

      if (this.bot.ws && this.bot.ws.readyState === this.bot.websocket.OPEN) {
        this.bot.ws.send(JSON.stringify(payload), (error) => {
          if (error) {
            console.error("Error sending EmoteRequest:".red, error);
            throw new HighriseError("Error sending EmoteRequest:".red)
          }
        });
      }
    } catch (error) {
      console.error(error); // Log the error message
    }
  };

  async react(user_id, reaction) {
    try {
      if (!reaction || typeof reaction !== 'string') {
        throw new HighriseError('Invalid Reaction. Please provide a valid string for reaction.');
      }

      if (!user_id || typeof user_id !== 'string') {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }

      const availableReactions = ['clap', 'heart', 'thumbs', 'wave', 'wink'];
      if (reaction && !availableReactions.includes(reaction)) {
        return console.error(`Invalid reaction '${reaction}'`.red + `\n` + `Available Reactions:`.green + `\n` + `${availableReactions.join("\n")}`);
      }
      const reactionRequest = new ReactionRequest(user_id, reaction, this.rid);

      const payload = {
        _type: 'ReactionRequest',
        reaction: reactionRequest.reaction,
        target_user_id: reactionRequest.target_user_id,
        rid: reactionRequest.rid
      };

      if (this.bot.ws.readyState === this.bot.websocket.OPEN) {
        this.bot.ws.send(JSON.stringify(payload), (error) => {
          if (error) {
            throw new HighriseError("Error sending EmoteRequest:".red)
          }
        });
      }
    } catch (error) {
      console.error(error); // Log the error message
    }
  };

  async tip(user_id, amount) {
    try {

      const BARS = {
        1: 'gold_bar_1',
        5: 'gold_bar_5',
        10: 'gold_bar_10',
        50: 'gold_bar_50',
        100: 'gold_bar_100',
        500: 'gold_bar_500',
        1000: 'gold_bar_1k',
        5000: 'gold_bar_5000',
        10000: 'gold_bar_10k'
      }


      if (!user_id || typeof user_id !== 'string') {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }

      if (typeof amount !== 'number' || !BARS.hasOwnProperty(amount)) {
        const availableAmounts = Object.keys(BARS).join(', ');
        throw new HighriseError(`Invalid amount. Please provide a valid amount from the available gold bars: ${availableAmounts}`.red);
      }

      if (this.bot.ws && this.bot.ws.readyState === this.bot.websocket.OPEN) {
        const tipUserRequest = new TipUserRequest(user_id, BARS[amount], this.rid);
        const payload = {
          _type: 'TipUserRequest',
          user_id: tipUserRequest.user_id,
          gold_bar: tipUserRequest.gold_bar,
          rid: tipUserRequest.rid
        }

        const sender = new SendPayloadAndGetResponse(this.bot);

        const response = await sender.sendPayloadAndGetResponse(
          payload,
          TipUserRequest.Response
        );

        return response.result.result
      }

    } catch (error) {
      console.error(error); // Log the error message
    }
  }

  transport(user_id, room_id) {
    try {
      if (!user_id || user_id === undefined || user_id === null) {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }
      if (!room_id || typeof room_id !== 'string') {
        throw new HighriseError(`Invalid room ID. Please provide a valid value for room ID.`.red);
      }

      const transportRequest = new MoveUserToRoomRequest(user_id, room_id, this.rid);
      const request = {
        _type: 'MoveUserToRoomRequest',
        user_id: transportRequest.user_id,
        room_id: transportRequest.room_id,
        rid: transportRequest.rid
      };

      if (this.bot.ws && this.bot.ws.readyState === this.bot.websocket.OPEN) {
        this.bot.ws.send(JSON.stringify(request), (error) => {
          if (error) {
            console.error('Error sending transport request:'.red, error);
            throw new HighriseError("Error sending TransportRequest:".red);
          }
        });
      }
    } catch (error) {
      console.error(error)
    }
  };

  async teleport(user_id, x, y, z, facing = 'FrontRight') {
    try {
      const validFacing = ['BackLeft', 'BackRight', 'FrontLeft', 'FrontRight'];

      if (!user_id || user_id === undefined || user_id === null) {
        throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
      }

      if (x === undefined || x === null || y === undefined || y === null || z === undefined || z === null) {
        throw new HighriseError('Invalid coordinates. Please provide valid values for x, y, and z.'.red);
      }

      if (typeof facing !== 'string') {
        throw new HighriseError('Invalid facing parameter. It must be a string.'.red);
      }

      if (!validFacing.includes(facing)) {
        throw new HighriseError(`Invalid facing parameter (${facing})`.red + `\n` + 'Available Facing Options:'.green + '\n' + `${validFacing.join("\n")}`);
      }

      const dest = new Position(x, y, z, facing);

      const teleportRequest = new TeleportRequest(user_id, dest, this.rid);
      const request = {
        _type: 'TeleportRequest',
        user_id: teleportRequest.user_id,
        destination: teleportRequest.destination,
        rid: teleportRequest.rid
      };

      if (this.bot.ws && this.bot.ws.readyState === this.bot.websocket.OPEN) {
        this.bot.ws.send(JSON.stringify(request), (error) => {
          const userData = CachedRoomUsers.get(request.user_id);
          if (userData) {
            userData.position = request.destination;
          }
          const botInfo = getBotInfo();
          if (botInfo.id === user_id) {
            const bot_position = {
              x: dest.x,
              y: dest.y,
              z: dest.z,
              facing: dest.facing
            }
            updateBotPosition(bot_position);
          }
          if (error) {
            console.error('Error sending teleport request:'.red, error);
            throw new HighriseError("Error sending TeleportRequest:".red);
          }
        });
      }
    } catch (error) {
      console.error(error);
    }
  };

  permissions = {
    get: async (user_id) => {
      try {

        if (!user_id || typeof user_id !== 'string') {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }
        if (this.bot.ws && this.bot.ws.readyState === this.bot.websocket.OPEN) {
          const getRoomPrivilegeRequest = new GetRoomPrivilegeRequest(user_id, this.rid);
          const payload = {
            _type: "GetRoomPrivilegeRequest",
            user_id: getRoomPrivilegeRequest.user_id,
            rid: getRoomPrivilegeRequest.rid
          };

          const sender = new SendPayloadAndGetResponse(this.bot); // Create an instance of SendPayloadAndGetResponse
          const response = await sender.sendPayloadAndGetResponse(
            payload,
            GetRoomPrivilegeRequest.Response
          );

          return response.content.content;
        }
      } catch (error) {
        throw new HighriseError("Error fetching users permissions:", error);
      }
    },

    moderator: {
      add: async (user_id) => {
        if (!user_id || typeof user_id !== 'string') {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }
        const permissions = { moderator: true };
        await this.change(user_id, permissions);
      },
      remove: async (user_id) => {
        if (!user_id || typeof user_id !== 'string') {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }
        const permissions = { moderator: false };
        await this.change(user_id, permissions);
      }
    },

    designer: {
      add: async (user_id) => {
        if (!user_id || typeof user_id !== 'string') {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }
        const permissions = { designer: true };
        await this.change(user_id, permissions);
      },
      remove: async (user_id) => {
        if (!user_id || typeof user_id !== 'string') {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }
        const permissions = { designer: false };
        await this.change(user_id, permissions);
      }
    }
  }

  async change(user_id, permissions) {
    try {
      const payload = {
        _type: 'ChangeRoomPrivilegeRequest',
        user_id: user_id,
        permissions: permissions,
        rid: this.rid
      };

      if (this.bot.ws.readyState === this.bot.websocket.OPEN) {
        this.bot.ws.send(JSON.stringify(payload), (error) => {
          if (error) {
            throw new HighriseError("Error sending ChangeRoomPrivilegeRequest:".red);
          }
        });
      };
    } catch (error) {
      console.error(error);
    }
  };

  backpack = {
    get: async (user_id) => {
      try {

        if (!user_id || user_id === undefined || user_id === null) {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }

        if (this.bot.ws && this.bot.ws.readyState === this.bot.websocket.OPEN) {
          const getBackpackRequest = new GetBackpackRequest(user_id, this.rid);
          const payload = {
            _type: "GetBackpackRequest",
            user_id: getBackpackRequest.user_id,
            rid: getBackpackRequest.rid
          };

          const sender = new SendPayloadAndGetResponse(this.bot); // Create an instance of SendPayloadAndGetResponse
          const response = await sender.sendPayloadAndGetResponse(
            payload,
            GetBackpackRequest.Response
          );

          return response.backpack.backpack;
        }

      } catch (error) {
        throw new HighriseError("Error fetching users backpack:", error);
      }
    }
  }
  outfit = {
    get: async (user_id) => {
      try {

        if (!user_id || typeof user_id !== 'string') {
          throw new HighriseError('Invalid user_id. Please provide a valid user_id'.red);
        }

        if (this.bot.ws && this.bot.ws.readyState === this.bot.websocket.OPEN) {
          const getUserOutfitRequest = new GetUserOutfitRequest(user_id, this.rid);
          const payload = {
            _type: "GetUserOutfitRequest",
            user_id: getUserOutfitRequest.user_id,
            rid: getUserOutfitRequest.rid
          };

          const sender = new SendPayloadAndGetResponse(this.bot); // Create an instance of SendPayloadAndGetResponse
          const response = await sender.sendPayloadAndGetResponse(
            payload,
            GetUserOutfitRequest.Response
          );

          return response.outfit.outfit

        }
      } catch (error) {
        throw new HighriseError("Error fetching users outfit:", error);
      }
    }
  }
};

module.exports = { Users };